CAMEL-10685: Added camel-cdi-jee component

RasPelikan opened the pull request:
The "transacted" configuration requires a Spring environment. We use JavaEE and we needed to use "transacted". We ported the Spring implementation of the TransactionErrorHandler to JavaEE.

There is only one major difference: Spring's TransactionErrorHandler derives from org.apache.camel.processor.RedeliveryErrorHandler which brings the behavior that redelivery is done within the transaction. In our implementation there is a RedeliveryErrorHandler which uses the TransactionErrorHandler to force a rollback and create a new transaction for each redelivery.

astefanutti commented:
@RasPelikan thanks for this PR. That looks very promising! I'll review it ASAP (unless somebody else does it before).

RasPelikan commented:
@astefanutti: Thank you. I recently updated (https://github.com/apache/camel/pull/1390/commits/2c3ba4dfbd58cf108ee9d0eb08023665d2d0819d) because the original exception was suppressed if redelivery exhausts.

oscerd commented:
Go for it @astefanutti :-)

davsclaus commented:
@astefanutti will you get some time in start/mid february to review this?

astefanutti commented:
@davsclaus I should be able to work on this mid February hopefully!

astefanutti commented on components/camel-cdi-jee/src/main/java/org/apache/camel/cdi/transaction/JavaEETransactionErrorHandlerBuilder.java:
```diff
@@ -0,0 +1,173 @@
+        return "TransactionErrorHandlerBuilder";
+    }
+
+    public boolean isWrapped() {
```
Is that used? it may be an internal.

astefanutti commented on components/camel-cdi-jee/src/main/java/org/apache/camel/cdi/transaction/RedeliveryErrorHandler.java:
```diff
@@ -0,0 +1,48 @@
+ * done within the transaction which is not appropriate in JavaEE since every
+ * error breaks the current transaction.
+ */
+public class RedeliveryErrorHandler extends org.apache.camel.processor.RedeliveryErrorHandler {
```
We may want to follow the current naming convention

astefanutti commented on components/camel-cdi-jee/src/main/java/org/apache/camel/cdi/transaction/SupportsEETransactionPolicy.java:
```diff
@@ -0,0 +1,15 @@
+import javax.inject.Named;
+
+@Named("PROPAGATION_SUPPORTS")
+public class SupportsEETransactionPolicy extends TransactionalJavaEETransactionPolicy {
```
Discrepancy with the naming convention

astefanutti commented on components/camel-cdi-jee/src/main/java/org/apache/camel/cdi/transaction/JavaEETransactionPolicy.java:
```diff
@@ -0,0 +1,136 @@
+        void run() throws Throwable;
+    }
+
+    @Resource(lookup = "java:/TransactionManager")
```
We may want to refine the strategy to lookup for the transaction manager. In a _micro service_ (or MicroProfile), we may want to use CDI and JTA only. So the lookup will be done through CDI and not JNDI. That being said, we can refine later on.

astefanutti commented on components/camel-cdi-jee/src/main/java/org/apache/camel/cdi/transaction/TransactionErrorHandler.java:
```diff
@@ -0,0 +1,370 @@
+ * within the transaction which is not appropriate in JavaEE since every error
+ * breaks the current transaction.
+ */
+public class TransactionErrorHandler extends ErrorHandlerSupport
```
I wonder how much that'd make sense to factorise with Spring implementation. Just a comment, we can refactor later.

davsclaus commented:
@RasPelikan do you have time to work on the feedback from @astefanutti - otherwise we may not be able to get this into the 2.19 release which would be a shame

RasPelikan commented:
I'm sorry, I tried to do the changes in our current sprint but I did not make it. 

Since a lot of time is gone since I made that pull request my situation in work has changed and we are within the ramp up of the next project. Next week I'm on vacation but I addressed this topic in our planing. The next sprint starts after my vacation and will include a story for getting this done. 

I would appreciate seeing those changes in 2.19 but at the moment I see no chance to get this done earlier.

davsclaus commented:
Thanks @RasPelikan for the update.

The 2.19.0 release has an ETA in about 4-5 weeks (eg mid towards end of April). So hopefully that should be within your sprint time.

davsclaus commented:
Hi 

Are you able to work on this in the near future, as we will start the release in week 16, eg in about 10 or so days

astefanutti commented:
@RasPelikan if you feel like you won't be able to work on this any time soon, I can propose I take your great PR as is and make the proposed changes above on top of it. Anyway, let us know what's the most comfortable way for you to move forward on this.

RasPelikan commented:
Hi,

I answered to davsclaus by mail but it seems that didn't work. What I wrote:

Hello,
I did the changes two days ago. I need to test the changes as part of our projects using this feature to ensure that still everything works fine. I really do my very best but I still cannot promise to be ready in time.
Stephan

In these days a lot is going on in office. Please give me the next days. I will respond as soon as possible. Thank you for your patience!

@astefanutti: you can review the changes in the meantime. I will push them tomorrow.

RasPelikan commented:
What I did:

1. I moved the package into the camel-cdi module and added JTA as an optional dependency.

1. I removed JavaEECamelContext and added "autoStartContexts" methods to CdiCamelConfiguration.

1. The package split was necessary because I used the Vetoed annotation in JavaEECamelContext. But both situations (two modules and JavaEECamelContext) are gone.

1. I removed the method "rollbackLoggingLevel" from JavaEERouteBuilder which was copied from the Spring implemention.

1. I removed the unused "wrapped" property

1. I removed the entire JavaEETransactionErrorHandlerBuilder constructor since it is not necessary any more. I think the "super()" call was also copied from Spring implemention, but I'm not sure.

1. Added "Override" annotation to method "supportTransacted()".

1. Fixed copy/paste issues.

1. Renamed class "org.apache.camel.cdi.transaction.RedeliveryErrorHandler" to "org.apache.camel.cdi.transaction.JavaEETransactionErrorHandler"

What I didn't change:

1. Unfortunately I don't know about XML support in CDI. If changes are necessary you have to apply them.

1. Changed further namings because of my lack of knowledge about valid naming conventions.

1. Removed the JNDI lookup for "java:/TransactionManager". This is intentionally because the TransactionManager is not something cannot be injected by common applications. The TransactionManager is something provided by application containers to support libraries like Camel to hook into the transaction lifecycle without the need for knowledge about the internal implementation of the container. 

I hope you will be pleased with my changes. At the moment I'm working on the integration into our projects to ensure everything still works as it should.

astefanutti commented on components/camel-cdi-jee/.gitignore:
```diff
@@ -0,0 +1,4 @@
+.classpath
```
I think that file can be removed.

astefanutti commented on components/camel-cdi-jee/pom.xml:
```diff
@@ -0,0 +1,65 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
I think that file can be removed.

astefanutti commented on components/camel-cdi-jee/src/main/resources/META-INF/beans.xml:
```diff
@@ -0,0 +1,3 @@
+<beans xmlns="http://java.sun.com/xml/ns/javaee" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://java.sun.com/xml/ns/javaee http://java.sun.com/xml/ns/javaee/beans_1_0.xsd">
```
I think that file can be removed.

astefanutti commented on components/camel-cdi/pom.xml:
```diff
@@ -115,6 +115,14 @@
       <optional>true</optional>
     </dependency>
 
+	<dependency>
```
Mind the formatting of POM files with spaces.

astefanutti commented on components/camel-cdi/src/main/java/org/apache/camel/cdi/CdiCamelExtension.java:
```diff
@@ -415,16 +415,18 @@ private void afterDeploymentValidation(@Observes AfterDeploymentValidation adv,
-                adv.addDeploymentProblem(exception);
-            }
+        if (configuration.autoStartContexts()) {
+	        for (CamelContext context : contexts) {
```
Mind the formatting.

astefanutti commented on components/camel-cdi/src/main/java/org/apache/camel/cdi/transaction/JavaEETransactionErrorHandler.java:
```diff
@@ -0,0 +1,48 @@
+ * done within the transaction which is not appropriate in JavaEE since every
+ * error breaks the current transaction.
+ */
+public class JavaEETransactionErrorHandler extends org.apache.camel.processor.RedeliveryErrorHandler {
```
FQN could be imported.

astefanutti commented on components/camel-cdi/src/main/java/org/apache/camel/cdi/transaction/TransactedDefinition.java:
```diff
@@ -0,0 +1,18 @@
+ * {@link JavaEETransactionErrorHandlerBuilder} to resolve configured policy
+ * references.
+ */
+public class TransactedDefinition extends org.apache.camel.model.TransactedDefinition {
```
I would rename it `JtaTransactedDefinition` and remove the parent FQN.

astefanutti commented on components/camel-cdi/src/main/java/org/apache/camel/cdi/JavaEERouteBuilder.java:
```diff
@@ -0,0 +1,24 @@
+ *
+ * @version
+ */
+public abstract class JavaEERouteBuilder extends RouteBuilder {
```
I would suggest to change every `JavaEE` suffixes to `Jta` to be more specific.

astefanutti commented on components/camel-cdi/src/main/java/org/apache/camel/cdi/transaction/TransactionalJavaEETransactionPolicy.java:
```diff
@@ -0,0 +1,121 @@
+/**
+ * Helper methods for transaction handling
+ */
+public abstract class TransactionalJavaEETransactionPolicy extends JavaEETransactionPolicy {
```
I wonder to what extent that'd make sense to have that class merged with `JavaEETransactionPolicy` and have it renamed `JtaTransactionPolicy`. Indeed, I would not expect an non transactional JTA transaction policy.

astefanutti commented:
@RasPelikan thanks a lot for the detailed summary. It looks great.

> Unfortunately I don't know about XML support in CDI. If changes are necessary you have to apply them.

No worries, I'll do the required changes as an enhancement.

> Changed further namings because of my lack of knowledge about valid naming conventions.

I understand. In the last review, I've just added an item to change the `JavaEE` suffixes to `Jta` to be more specific. Let me know if that's OK with you.

> Removed the JNDI lookup for "java:/TransactionManager". This is intentionally because the TransactionManager is not something cannot be injected by common applications. The TransactionManager is something provided by application containers to support libraries like Camel to hook into the transaction lifecycle without the need for knowledge about the internal implementation of the container.

Sure. I'll add the fallback to CDI lookup support as a subsequent enhancement.

I've just submitted a last round of review. Once you're done with them, would you mind squashing the commits as that would help keep the history clean.

Thanks again for that great contribution!

RasPelikan commented:
I renamed every occurence of JavaEE to Jta.

I built Camel and used that version in one of my projects using the transaction feature. The changes within the project were straight forward. For configuration I added

```
    public void configuration(@Observes CdiCamelConfiguration configuration) {
        configuration.autoStartContexts(false);
    }

```
to my CamelContextProducer. This works fine. The context is not started until my CamelBooter bean (annotated by @Singleton @ApplicationScoped @Startup) starts the context.

That being said I can report that using the new version works as good as my local version worked. I think we are save to clear that feature.

RasPelikan commented:
I've forgot to say that I do not mind squashing the history. I think you can do that on merging the pull request - there is no need to take any action by me, isn't it?

astefanutti commented:
Great! Sure I can squash the commits while merging. I may restore the `org.apache.camel.cdi.transaction` package instead of `org.apache.camel.cdi.jpa` if that's fine for you. I'd rather go for a _functional_ convention for the package name... I'll do the merge very soon.

RasPelikan commented:
>  I may restore the package

Sure. I wasn't sure about the naming. Of course it's fine.

>I'll do the merge very soon.

I'm looking forward to that version. Thanks for being patient about me.

astefanutti commented:
> I'm looking forward to that version. Thanks for being patient about me.

Thanks a lot for that very valuable PR! It's been a pleasure iterating with you on it 😉.

RasPelikan commented:
you're welcome!

astefanutti commented:
I've just merged it @8092e89f63515889bbd26457d1380b5c6cf68eb1.

@RasPelikan thanks a lot for the great work and your patience! I've polished it a bit in subsequent commits. I'll complement it (XML DSL support) ASAP.

@davsclaus @oscerd you can close that PR (I don't have the rights).

oscerd commented:
@astefanutti Thanks for merging.

@RasPelikan can you close the PR? Thanks for the contribution
